### DISCUSSION TITLE
Announcing v4.0.0

### DISCUSSION THREAD

**Discussion by nick-thompson** [elemaudio/elementary#58] [Announcements]
### New Features
* Support for multi-channel graph nodes
    * With new mc node types
        * el.mc.sampleseq
        * el.mc.sampleseq2
        * el.mc.table
        * el.mc.capture
        * el.mc.sample
* Include active/passive indication in GraphNode process calls
* Expose root node fade times for custom Renderer implementations

### Improvements
* Rewrite frontend library with explicit props args
* Rewrite SharedResource with an explicit underlying f32 type
* Rewrite GC implementation 
* Build wasm binaries with an underlying f64 type
* Add safeguards to activateRoots and optimize render ops

### Bugfixes
* Bugfixes and improvements in el.sampleseq
* Fix capture event type
* Pre-emptively zero output buffers in wasm rendering to prevent garbage
* Fix metro node type def
* Bugfix in el.adsr for zero-valued arguments

## New Contributors
* @cannc4 and @ncblair made their first contribution in https://github.com/elemaudio/elementary/pull/41
* @awood314 made their first contribution in https://github.com/elemaudio/elementary/pull/48 and a second in https://github.com/elemaudio/elementary/pull/52
* @d4rekanguok made their first contribution in https://github.com/elemaudio/elementary/pull/56
* @b00lduck made their first contribution in https://github.com/elemaudio/elementary/pull/50

**Full Changelog**: https://github.com/elemaudio/elementary/compare/@elemaudio/core@3.2.1...@elemaudio/core@4.0.0

<hr /><em>This discussion was created from the release <a href='https://github.com/elemaudio/elementary/releases/tag/%40elemaudio/core%404.0.0'>@elemaudio/core@4.0.0</a>.</em>

**Comment by davay42** [comment]
Congratulations on this big release!

I've tried using this version with my Elements synth here https://github.com/chromatone/elements/ but I keep getting errors from the el.noise() function about `a` being undefined. I can't understand what this `a` stands for and how is it defined.

<img width="430" alt="Screenshot 2024-12-09 at 08 55 10" src="https://github.com/user-attachments/assets/7b0f2239-d8c8-464b-a609-40cefb4c11ff">

<img width="327" alt="Screenshot 2024-12-09 at 08 56 13" src="https://github.com/user-attachments/assets/0c71dd6d-6c8d-4462-a09a-904a13373b4a">

**Reply by nick-thompson** [reply]
Thank you @davay42! That was a bug that slipped through in the release. I've just shipped a patch 4.0.1, please try that and see if your issue resolves?

**Reply by davay42** [reply]
Yes! Tried the new version - all works! Will update my other Elementary projects to the new version soon!

**Reply by nick-thompson** [reply]
Great!
